### PR TITLE
WIP - Refactoring analyzer tests to use custom docker registry

### PR DESCRIPTION
**Pull request by jjbustamante**
### References

- This Pull Request is created after [PR #646](https://github.com/buildpacks/lifecycle/pull/646) was merged

### Context

After merging  [PR #646](https://github.com/buildpacks/lifecycle/pull/646) we agreed to refactor the analyzer tests and simplifying the number of docker registries we are using. 

### Description

- The PR removes the `authRegistry` and the `readOnlyRegistry` and replaces them with just 1 `testRegistry` that uses the `ImagePrivilges` approach implemented in  imgUtil repo [PR #126](https://github.com/buildpacks/imgutil/pull/126) to configure the registry to handle the required behaviors.

### Depends on

- This PR depends on imgUtil repo [PR #126](https://github.com/buildpacks/imgutil/pull/126)

### CODE REVIEW

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:51]: Is this an accurate comment? Don't we need read/write for cache?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:156]: Don't we still need this?

Comment (natalieparellano) [reply acceptance/analyzer_test.go:156]: Oh I see, now it's being set here: https://github.com/buildpacks/lifecycle/pull/685/files#diff-c4e4b4a55ebbd71443c5260020b287f3018cb7bb4c65c79e74554b7f97c45656R155

Comment (natalieparellano) [reply acceptance/analyzer_test.go:156]: IIUC, since we only have one registry now, we need to actually create the read only image instead of relying on access through the other registry to give the correct access permissions.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:165]: Can we keep the random string?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:174]: I'm confused how this is working. The tests should fail if the cache image is read only.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:180]: Why is this needed?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:966]: This would already be accessible because of line 962

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:1135]: I'm a bit confused how this is working. The `inaccessibleImage` doesn't have any privileges defined on it, so it should fall through to the Basic Auth - but then we provided valid credentials in `fixtures.authRegAuthConfig`, so how is it failing to ensure the access?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [reply acceptance/analyzer_test.go:180]: (it seems like a no-op, except for setting image name on `fixtures`)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:165]: Yeap

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:174]: There are a couple of tests actually verifying the read/write error is thrown when a read-only cache image is used. But, there are already two test cases using `fixtures.readOnlyRegCacheImage`, remember I am not adding or modifying the test cases, I am just changing the reference to the image they are using. Those test cases are validating if `cache metadata is restore`

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:180]: Yes, this one is just creating an image without reading or writing access that is being used in some test cases

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:1135]: Well, with the change proposed in imgUtil [PR 126](https://github.com/buildpacks/imgutil/pull/126) this image can help to understand it

wrapping 2

- The request to the `inaccessibleImage` fails with 401 Unauthorized, because the first filter to handle any registry request is the `Custom` and it fills the conditions (it is on permissions map), it never goes into the `Basic Auth` filter, that's why the validation fails correctly.

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:51]: Mmm, you are right, it shouldn't be read-only. I will fix it

Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/27c678fc98ac3d25a2f3c0727ebff90aa0640ba1)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/27c678fc98ac3d25a2f3c0727ebff90aa0640ba1)

File: acceptance/analyzer_test.go
Comment (natalieparellano) [reply acceptance/analyzer_test.go:180]: Do we need it? What happens if this is left out?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [reply acceptance/analyzer_test.go:1135]: > it is on permissions map

This helps answer my question [here](https://github.com/buildpacks/lifecycle/pull/685/files#r685552562). Thinking about how this could be made more clear...

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: We can remove them, but I will need to send a different image on some test cases, the ones where I am validating read or read/write access. I can try to do it, let me check

Comment (jjbustamante) [reply acceptance/analyzer_test.go]: I will need to create at least one image with only writable permissions, there is a test in line 960 where I need an image that is not readable, we can accomplish that with a only writable image or an image without any permissions like this one.

Comment (natalieparellano) [reply acceptance/analyzer_test.go]: I think I understand how this is working now (see https://github.com/buildpacks/lifecycle/pull/685#discussion_r686323993). This seems fine to me.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [reply acceptance/analyzer_test.go:960]: (nevermind)

File: acceptance/analyzer_test.go
Comment (yaelharel) [inline acceptance/analyzer_test.go:51]: I think the comments are redundant.

File: acceptance/analyzer_test.go
Comment (yaelharel) [inline acceptance/analyzer_test.go:43]: Maybe it's better to use the convention like in lines 49-51 and call this variable `readWriteAuthRegAppImage`.

Comment (yaelharel) [reply acceptance/analyzer_test.go:43]: Same comment apply to lines 45-46.

Comment (natalieparellano) [reply acceptance/analyzer_test.go:43]: I think we could remove the `AuthReg` (and `readOnlyReg`) part since now we have only one registry.

Comment (natalieparellano) [reply acceptance/analyzer_test.go:43]: So `readWriteAppImage` and `readOnlyAppImage`, etc.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:169]: It seems that this image has read & write permission... should we make another image for `readOnlyRegCacheImage`?

File: go.mod
Comment (aemengo) [inline go.mod:26]: Let's make sure that this doesn't make it in the final merge 🤓

Comment (jjbustamante) [reply go.mod:26]: Yeah! I will remove that once the imgUtil PR has been merged

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: I refactored the names in the last [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/59c111c24edd1c3bb2cdcaba6ba521a5b3718d4d)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: I did a refactoring following the advice, check the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/59c111c24edd1c3bb2cdcaba6ba521a5b3718d4d)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: check the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/59c111c24edd1c3bb2cdcaba6ba521a5b3718d4d)

File: go.mod
Comment (yaelharel) [inline go.mod:25]: Do we still need it?

File: go.sum
Comment (yaelharel) [inline go.sum:453]: Do we still need it?

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:168]:
```suggestion
	someReadOnlyCacheImage := "some-read-only-cache-image-" + h.RandString(10)
```

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:177]:
```suggestion
	testRegistry.SetReadOnly(someReadOnlyCacheImage)
```

File: go.mod
Comment (jjbustamante) [reply go.mod:25]: Not sure if the [PR #126](https://github.com/buildpacks/imgutil/pull/126) on imgUtil repo was already merged, if it was, I will delete those lines

File: go.sum
Comment (jjbustamante) [reply go.sum:453]: Not sure if the [PR #126](https://github.com/buildpacks/imgutil/pull/126) on imgUtil repo was already merged, if it was, I will delete those lines

Comment (yaelharel) [reply go.sum:453]: You're right. I thought it was merged yesterday.

File: go.mod
Comment (yaelharel) [reply go.mod:25]: You're right. I thought it was merged yesterday.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:51]:
```suggestion
	daemonAppImage         string
	daemonCacheImage       string
	inaccessibleImage      string
	readOnlyAppImage       string
	readOnlyCacheImage     string
	readWriteAppImage      string
	readWriteCacheImage    string
	readWriteOtherAppImage string
	regAuthConfig          string
```
(nit: sorted)

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:1221]:
```suggestion
func buildRegistryImage(t *testing.T, repoName, context string, registry *ih.DockerRegistry, buildArgs ...string) (string, string) {
```
I wonder if it still makes sense to have this function return the auth. Since we're only getting the auth one time... maybe we could just pull that out, and then we don't have to ignore the second return value every other time.

File: go.mod
Comment (natalieparellano) [reply go.mod:25]: I don't think so, the other PR was merged!

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:403]: Do we even need this variable? We could use `fixtures.readWriteAppImage` on line 414...

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:618]: This test is explicitly verifying that the credentials passed to the analyzer were handled correctly - thus I don't think we want to provide our hard-coded permissions that bypass basic auth! We probably need one more cache image that doesn't have permissions set.

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:643]: Same comment as https://github.com/buildpacks/lifecycle/pull/685/files#r691403233

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:793]: Same comment as https://github.com/buildpacks/lifecycle/pull/685/files#r691403233 (but for app)

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:816]: Same comment as https://github.com/buildpacks/lifecycle/pull/685/files#r691404307

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:996]: Same comment as https://github.com/buildpacks/lifecycle/pull/685/files#r691403233

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:1022]: Same comment as https://github.com/buildpacks/lifecycle/pull/685/files#r691403233

File: go.mod
Comment (yaelharel) [reply go.mod:25]: Yeah, as I wrote, my mistake.

**Comment by natalieparellano** [comment]
@jjbustamante this is looking good! There are a few places where we are explicitly testing the auth, so I don't think we want to pass images that are going to bypass basic auth for the registry. We'll probably need to create a couple extra fixtures that don't have permissions set.

**Comment by natalieparellano** [comment]
Also it looks like there is a merge conflict now :( I hope it won't be too tedious to resolve. I am happy to pair on it if you'd like (I think the PR that introduced the conflict was mine).

File: go.mod
Comment (jjbustamante) [reply go.mod:25]: @natalieparellano the imgutil is still not merged into main, the main was merged into the PR branch

Comment (natalieparellano) [reply go.mod:25]: Whoops! Done now.

Comment (yaelharel) [reply go.mod:25]: At least I'm not the only one that thought it was already merged yesterday 😃

File: go.mod
Comment (jjbustamante) [reply go.mod]: Removed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: go.sum
Comment (jjbustamante) [reply go.sum]: Removed at [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: I did the change, you can take a look and give me your thoughts

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: Fixed in the following [commit](https://github.com/buildpacks/lifecycle/pull/685/commits/7ce8bb8297e15af0aab609e3df4c462172972ef9)

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go]: I added a couple of fixtures without using any permissions and replace it in this test case

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:1257]: Ahh I'm sorry, I think I was not clear in my earlier comment. I meant remove `authConfig`, and keep `regRepoName`!

**Review by natalieparellano — APPROVED** [review]
Looks good! `buildAuthRegistryImage` may be a bit weird now (apologies for not being clear before). I'd be happy to change it in a "polish" PR if you wish.

File: acceptance/analyzer_test.go
Comment (jjbustamante) [reply acceptance/analyzer_test.go:1257]: Mmm... but  I think we need the **Credentials** created for the registry. right? it is used only once because it is the same registry, but it still a required data to fill the fixture

Comment (natalieparellano) [reply acceptance/analyzer_test.go:1373]: Ohh you're right. The thing is, `auth.BuildEnvVar` - it expects to be called with a repo, but the auth is **per registry**, not per repo. Probably we should refactor `auth.BuildEnvVar` to just expect a registry, but that's another story...

In any case, I think you could call lines 1370+1371 outside of this function with the first repo that is created.

**Comment by jjbustamante** [comment]
@natalieparellano I tried to fix the merging conflict, but now there are some tests failing, I think probably I missed something maybe we can try to check

File: acceptance/analyzer_test.go
Comment (natalieparellano) [inline acceptance/analyzer_test.go:255]:
```suggestion
	h.DockerImageRemove(t, fixtures.readWriteAppImage)
	h.DockerImageRemove(t, fixtures.readWriteCacheImage)
	h.DockerImageRemove(t, fixtures.readWriteOtherAppImage)
	h.DockerImageRemove(t, fixtures.someAppImage)
	h.DockerImageRemove(t, fixtures.someCacheImage)
```

Comment (jjbustamante) [reply acceptance/analyzer_test.go:255]: Done [here](https://github.com/buildpacks/lifecycle/pull/685/commits/05fdc83045bec339338093adc3b3ee2ab13d17d8)

**Pull request merged**
